src/pstore/pstore.c

```diff
@@ -0,0 +1,403 @@
+        /* Move files, reconstruct dmesg.txt */
+        PStoreEntry *pe;
+        char *dmesg = NULL;
+        size_t dmesg_size = 0;
+        _cleanup_free_ char *dmesg_id = NULL;
+
+        /* Handle each dmesg file: files processed in reverse
+         * order so as to properly reconstruct original dmesg */
+        for (size_t n = list->n_entries; n > 0; n--) {
+                bool move_file_and_continue = false;
+                _cleanup_free_ char *pe_id = NULL;
+                char *p;
+                size_t plen;
+
+                pe = &list->entries[n-1];
+
+                if (pe->handled)
+                        continue;
+                if (!startswith(pe->dirent.d_name, "dmesg-"))
+                        continue;
+
+                if (endswith(pe->dirent.d_name, ".enc.z")) /* indicates a problem */
+                        move_file_and_continue = true;
+                p = strrchr(pe->dirent.d_name, '-');
+                if (!p)
+                        move_file_and_continue = true;
+
+                if (move_file_and_continue) {
+                        /* A dmesg file on which we do NO additional processing */
+                        (void) move_file(pe, NULL);
+                        continue;
+                }
+
+                /* See if this file is one of a related group of files
+                 * in order to reconstruct dmesg */
+
+                /* When dmesg is written into pstore, it is done so
+                 * in small chunks, whatever the exchange buffer size is
+                 * with the underlying pstore backend (ie. EFI may be ~2KiB),
+                 * which means an example pstore with approximately 64KB of
+                 * storage may have up to roughly 32 dmesg files that could be
+                 * related, depending upon the size of the original dmesg.
+                 * Here we look at the dmesg filename and try to discern if
+                 * files are part of a related group, meaning the same original
+                 * dmesg.
+                 * The two known pstore backends are EFI and ERST. These backends
+                 * store data in the Common Platform Error Record, CPER, format.
+                 * The dmesg- filename contains the CPER record id, a 64bit number
+                 * (in decimal notation). In Linux, the record id is encoded with
+                 * two digits for the dmesg part (chunk) number and 3 digits for
+                 * the count number. So allowing an additional digit to compensate
+                 * for advancing time, this code ignores the last six digits of the
+                 * filename in determining the record id.
+                 * For the EFI backend, the record id encodes an id
+                 * in the upper 32 bits, and a timestamp in the lower 32-bits.
+                 * So ignoring the least significant 6 digits has proven to
+                 * generally identify related dmesg entries.  */
```

Comment:
Could you rewrap those paras? This varying width makes them harder to read.
Also, we usually an empty line to separate paragraphs usually.

Reply:
I've tried to make these fit at approximately 70 chars.